### PR TITLE
fix(DB/conditions): Prevent NPC multiple spawns

### PR DESCRIPTION
**Pull request by Kosiak**
## Changes Proposed:
- Add Condition to check if NPC is present already.

## Issues Addressed:

- Closes  [[Display error] Sha'ni Proudtusk (npc=9136) not despawning #3106](https://github.com/chromiecraft/chromiecraft/issues/3106)

## Tests Performed:
 Had quest [Dreadmaul Rock](https://classic.wowhead.com/quest=3821/dreadmaul-rock). Clicked on gameobject. Now only spawns one NPC wait for de-spawn and click again spawns again. Relog to different character wait for de-spawn and tried again.

**She will appear in ghost form if you have the quest.** Source : [wowwiki](https://wowwiki-archive.fandom.com/wiki/Sha%27ni_Proudtusk) notes.

## How to Test the Changes:

1. Accept quest [Dreadmaul Rock](https://classic.wowhead.com/quest=3821/dreadmaul-rock)
2. Click on required gameobject couple times
3. Should only spawn one npc.
4. Wait for npc to de-spawn. Repeat steps 2-3.
5. Relog to different character. Repeat steps 1-4.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1647050684059087200.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1647050684059087200');
+
+UPDATE `acore_world`.`smart_scripts` SET `event_flags` = 1 WHERE `entryorguid` = 160445;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1647050684059087200.sql:3]:
```suggestion
UPDATE `smart_scripts` SET `event_flags` = 1 WHERE `entryorguid` = 160445;

```

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Comment by Nefertumm** [comment]
I don't know about this. According to the smart_script it should despawn after 60 secs, if we set the flags to not repeatable, you won't be able to spawn it with another player.. Or the not repeatable flag is just for X player and it won't matter for player Y?

**Comment by Kosiak** [comment]
> I don't know about this. According to the smart_script it should despawn after 60 secs, if we set the flags to not repeatable, you won't be able to spawn it with another player.. Or the not repeatable flag is just for X player and it won't matter for player Y?

I will try this later today. And if its not spawning will try to find another way.

**Comment by Lynethris** [comment]
Tested and work.

**Comment by Kosiak** [comment]
OK. Thanks. I will look in to it today. And update pr.

**Comment by Kosiak** [comment]
@Nefertumm. Tried for hours different ways and I don't see any options how to do that trough database. The best solution trough DB would be to change remains **gameobject_template_addon.flags** to **GO_FLAG_NOT_SELECTABLE(16)** but i can't come up with idea how to remove it after... I thought about doing it on Shani Proudtusk de-spwan but I can't find that event in **smart_scripts**. Maybe someone will have more luck.

**Comment by Nefertumm** [comment]
Maybe try to despawn it before doing spawning it again

**Comment by Kosiak** [comment]
> Maybe try to despawn it before doing spawning it again
Would you elaborate? I found this [fix](https://github.com/TrinityCore/TrinityCore/issues/249) for this quest from 2011. It seems it doesnt change till now. And it was just a hack. I think we need a C script for spawning it right way but my C skill is stayed in 2011 :)

**Comment by Lynethris** [comment]
@Kosiak can I test it ?

**Comment by Kosiak** [comment]
> @Kosiak can I test it ?

Yes please.

File: src/server/scripts/World/go_scripts.cpp [hunk 2]
```diff
@@ -266,6 +267,31 @@ class go_resonite_cask : public GameObjectScript
...
+};
+
+class go_proudtusk_remains : public GameObjectScript
+{
+public:
+    go_proudtusk_remains() : GameObjectScript("go_proudtusk_remains") { }
+
+    bool OnGossipHello(Player* player, GameObject* go) override
+    {
+        // xinef: prevent spawning hundreds of them
```
Comment (Nefertumm) [inline src/server/scripts/World/go_scripts.cpp:287]: Remove xinef comment, not necessary :P

Comment (Kosiak) [reply src/server/scripts/World/go_scripts.cpp:287]: I thought its a bickname of the guy whos code i used to fix this. But if u say.

**Comment by Nefertumm** [comment]
Aaand change the title for a proper one, please

**Comment by Nyeriah** [comment]
Objection: you could've used conditions to prevent the SAI from running if another creature is nearby. We usually attempt to move cpp to SAI where possible so unless it absolutelly doesn't work we should still keep it in the database

**Comment by Kosiak** [comment]
> Thanks for the PR, but I have an objection: you could've used conditions to prevent the SAI from running if another creature is nearby. We usually attempt to move cpp to SAI where possible so unless it absolutelly doesn't work we should still keep it in the database

Agree. The issue is that we cant call creature_distance (75) event from gossip_hello(64) event as they are both event_type. And we cant link them together as we would need to use link(61) event after gossip_hello event. Tried to do it in db for many hours with no success.  And we reaaly should fix it as we dont want someone to spawn too many npc just for fun.

**Comment by Nyeriah** [comment]
> > Thanks for the PR, but I have an objection: you could've used conditions to prevent the SAI from running if another creature is nearby. We usually attempt to move cpp to SAI where possible so unless it absolutelly doesn't work we should still keep it in the database
> 
> Agree. The issue is that we cant call creature_distance (75) event from gossip_hello(64) event as they are both event_type. And we cant link them together as we would need to use link(61) event after gossip_hello event. Tried to do it in db for many hours with no success. And we reaaly should fix it as we dont want someone to spawn too many npc just for fun.

I'm not sure if I understood it. What do you mean they're both event_type? Also, why are you changing gameobject_template type? That field is never wrong :\

https://www.azerothcore.org/wiki/conditions

**Comment by Kosiak** [comment]
> > > Thanks for the PR, but I have an objection: you could've used conditions to prevent the SAI from running if another creature is nearby. We usually attempt to move cpp to SAI where possible so unless it absolutelly doesn't work we should still keep it in the database
> > 
> > 
> > Agree. The issue is that we cant call creature_distance (75) event from gossip_hello(64) event as they are both event_type. And we cant link them together as we would need to use link(61) event after gossip_hello event. Tried to do it in db for many hours with no success. And we reaaly should fix it as we dont want someone to spawn too many npc just for fun.
> 
> I'm not sure if I understood it. What do you mean they're both event_type? Also, why are you changing gameobject_template type? That field is never wrong :\
> 
> https://www.azerothcore.org/wiki/conditions

To start summoning script we need to interact with GO so we using on_gossip_hello after that we can do action. But there is no action to check for nearby npc. But there is event to check for newrby npc but we cant use this as it will be triggering non stop and we cant call this event from clicking on object. What for conditions. I tried all conditions whis checks for creatures but they need alive state. Our creature doesn't  exist which makes it neither dead or alive. Gameobject type is questgiver now but its not really a quest giver is it? If anyone can show a way how to do it trough DB im alll ears :)

**Comment by Kosiak** [comment]
@Nyeriah was right. It can be done trough database conditions. It can be tested now.

**Comment by Kitzunu** [comment]
You have deleted an entire File

**Comment by Kosiak** [comment]
> You have deleted an entire File

Lol sorry. I thought it would only delete changes I made. Hope it's ok now.

**Comment by Lynethris** [comment]
All work.

**Comment by Kitzunu** [comment]
no, no, no. We do not need conditions. We just need to fix the SAI to properly despawn the creature when the event is done. Making a condition to prevent it to spawn if there are other creatures nearby is just a hack.

**Comment by Kosiak** [comment]
> no, no, no. We do not need conditions. We just need to fix the SAI to properly despawn the creature when the event is done. Making a condition to prevent it to spawn if there are other creatures nearby is just a hack.

Yes but how to prevent peoples from spawning more then 1 npc? They despawn properly. Another optinion will be to disable GO when NPC nearby. But it will be same as this just different method.

**Comment by Kitzunu** [comment]
wdym? can 1 player click the gob multiple times?
If so, either make an action list to apply a flag to the gob so it cannot be interacted with for the duration until the npc despawns.

How does blizzard handle this quest?

**Comment by Kosiak** [comment]
> wdym? can 1 player click the gob multiple times? If so, either make an action list to apply a flag to the gob so it cannot be interacted with for the duration until the npc despawns.
> 
> How does blizzard handle this quest?

Yes 1 player can spawn unlimited amount of npc. Each npc despawns after 60 sec. Its hard to determine how blizzard handles it without sniff. Only 1 video i found shows that corpses stays still active.  But no videos when someone tries to click it twice :d So we need either this option to check newrby npc or option when we disable GOB. Either way it's a hack i think.

**Comment by Kitzunu** [comment]
Let me see if I can find sniffs

**Comment by Kitzunu** [comment]
Okay I have a sniff now, but cannot parse it atm. But what I have been told is that you do not need to click the Gob, but that the creature should spawn once you are close enough to the Gob

**Comment by Kosiak** [comment]
> Okay I have a sniff now, but cannot parse it atm. But what I have been told is that you do not need to click the Gob, but that the creature should spawn once you are close enough to the Gob

Intresting. Wowhead comments says that u need to click GOB. (Not 3.3.5 patch though) very old and new classic comments. And all videos shows that u click on GOB. Gonna wait to see whats in that sniff of yours.

**Comment by neifion-00000000** [comment]
https://www.youtube.com/watch?v=37BSM5KUrsw  (Video title unrelated to conversation going on here so please don't take it to heart unless you're ddraig) 

This shows the proper functionality (and the blizzlike state of disaster for pathing of the quest givers as a bonus).  I have also donated a sniff but I may have clicked the gameobject during that sniff, I don't recall (it's an old sniff)

**Comment by Kosiak** [comment]
> https://www.youtube.com/watch?v=37BSM5KUrsw (Video title unrelated to conversation going on here so please don't take it to heart unless you're ddraig)
> 
> This shows the proper functionality (and the blizzlike state of disaster for pathing of the quest givers as a bonus). I have also donated a sniff but I may have clicked the gameobject during that sniff, I don't recall (it's an old sniff--we are unable to parse current version) but regardless, interaction with the gameobject is not necessary
> 
> In general, wowhead comments are not to be trusted especially since we have a better source available. While they have some value if we are completely missing functionality, past that and on almost every issue they come up in, they seem to prove their unreliability.

Not everybody got access to sniffs unfortunately. Would make job easier. But its weird to say we should not trust wowhead comments. Every third pr someone says dont trust this or that site. We will run out of sites to trust soon. :D @Kitzunu are we agree that it should be like in video? Spawn creature on player nearby and has quest? And does she spawn or changes state from invisible to visible?

**Comment by Kosiak** [comment]
> 

My WoW times passed long ago, don't have time to play anymore. I understand your problem with wowhead comments, but as I pointed before u'r sniff party is private and only so many peoples get access to sniffs(understandably trough). So me or other contributors have no other choice to search for info online(wowhead, wowwiki, wowpedia). And issues doesn't seem to finish anytime soon as there is just not enough sniffers + some things breaks when someone changes things. Even now I understand that she appears when player nearby but still don't have any info how long she stays, is she spawns or just comes out of invisibility and I wont find this info anywhere, I can fix it either way (spawn or appear) but someone will be not happy again as it's not 100% blizzlike. Some other things can be broken because of 1 line of code changed in core but no one checked if it broke something and no one asks for proof.

**Comment by neifion-00000000** [comment]
> My WoW times passed long ago, don't have time to play anymore. I understand your problem with wowhead comments, but as I pointed before u'r sniff party is private and only so many peoples get access to sniffs(understandably trough). So me or other contributors have no other choice to search for info online(wowhead, wowwiki, wowpedia). And issues doesn't seem to finish anytime soon as there is just not enough sniffers + some things breaks when someone changes things. Even now I understand that she appears when player nearby but still don't have any info how long she stays, is she spawns or just comes out of invisibility and I wont find this info anywhere, I can fix it either way (spawn or appear) but someone will be not happy again as it's not 100% blizzlike. Some other things can be broken because of 1 line of code changed in core but no one checked if it broke something and no one asks for proof.

While I realize its not a perfect solution, we do respond to requests for sniffs on discord.  We can remove the private info from the parses but it has to be done manually on each parse (and parsing is done on demand because there are semi-frequent updates to the parser so keeping old parses isn't that useful)

**Comment by Kosiak** [comment]
> > My WoW times passed long ago, don't have time to play anymore. I understand your problem with wowhead comments, but as I pointed before u'r sniff party is private and only so many peoples get access to sniffs(understandably trough). So me or other contributors have no other choice to search for info online(wowhead, wowwiki, wowpedia). And issues doesn't seem to finish anytime soon as there is just not enough sniffers + some things breaks when someone changes things. Even now I understand that she appears when player nearby but still don't have any info how long she stays, is she spawns or just comes out of invisibility and I wont find this info anywhere, I can fix it either way (spawn or appear) but someone will be not happy again as it's not 100% blizzlike. Some other things can be broken because of 1 line of code changed in core but no one checked if it broke something and no one asks for proof.
> 
> While I realize its not a perfect solution, we do respond to requests for sniffs on discord. We can remove the private info from the parses but it has to be done manually on each parse (and parsing is done on demand because there are semi-frequent updates to the parser so keeping old parses isn't that useful)

Ok I didn't know we can ask for parses I thought it's only limited to some peoples. So I will try to fix it to be like in the video, on player near GOB.

**Comment by Kosiak** [comment]
Update: I tried to use SMART_EVENT_NEAR_PLAYERS on GOB but apparently u can't use it with GOB. Only creatures. So either smart scripts need to be redone or there should be invisible trigger npc. At this point Im lost which path to take. it's easy to do it trough cpp but SAI is just will be pain in ass if im not missing something. I will look trough parse as soon as i got time, but if anyone got ideas please share.

File: data/sql/updates/pending_db_world/rev_1647050684059087200.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1647050684059087200');
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1647050684059087200.sql:1]: delete this

**Comment by acidmanifesto** [comment]
What is the current status of this? Is it abandoned?

**Comment by acidmanifesto** [comment]
@Kosiak can you fetch the upstream of this pr to bring it current with the current master so we can get a fresh re-run of the work flow to see if we still are getting those dry run failures?

**Pull request closed without merging**